File: #5. 모의고사 1회/장효빈/problem2_체육대회.py
# 풀이 시간: 10분
# 시간 복잡도: O(N!)
# 코드 설명: 각 종목에 선발할 학생들의 번호의 경우의 수를 순열로 구함.
#           각 경우의 수마다 계산해서 최대값을 구함.
# 취약한 부분: 

from itertools import permutations

def solution(ability):
    n = len(ability[0])
    
    perm = list(permutations(range(len(ability)), n))
    
    answer = 0
    
    for p in perm:
        total = 0
        for i in range(n):
            total += ability[p[i]][i]
        answer = max(answer, total)
    
    return answer

File: #5. 모의고사 1회/박정현/problem2_체육대회.java
// 풀이 시간: 60분
// 시간 복잡도: O(N^2)?? 순열.... 복잡도.. ?
// 코드 설명: 순열을 돌려서 모든 경우를 다 찾아봄.
// 취약한 부분: 시간적으로 최악일 것.. 하지만 이게 되네? 였다...

class Solution {
    int size, max;
    int[] arr;
    int[][] aability;
    boolean[] visited;
    public int solution(int[][] ability) {
        size = ability[0].length;
        arr = new int[size];
        visited = new boolean[ability.length];
        aability = ability;
        int answer = 0;
        
        perm(0);
        
        return max;
    }
    
    public void perm(int sidx) {
        if(sidx == size) {
            int temp = 0;
            for(int i=0; i<size; i++) {
                temp += aability[arr[i]][i];
            }
            
            max = Math.max(max, temp);
            return;
        }
        
        for(int i=0; i<aability.length; i++) {
            if(!visited[i]) {
                visited[i] = true;
                arr[sidx] = i;
                perm(sidx+1);
                visited[i] = false;
            }
        }
    }
}
